test/unit/rh_cloud_http_proxy_test.rb

```diff
@@ -0,0 +1,48 @@
+require 'test_plugin_helper'
+
+class RhCloudHttpProxyTest < ActiveSupport::TestCase
+  setup do
+  end
+
+  test 'Proxy selected in the correct order' do
+    global_content_proxy_mock = 'http://global:content@localhost:8888'
+    katello_cdn_proxy_mock = 'http://katello:cdn@localhost:8888'
+    global_foreman_proxy_mock = 'http://global:foreman@localhost:8888'
+
+    FactoryBot.create(:setting, :name => 'http_proxy', :value => global_foreman_proxy_mock)
+
+    original_cdn_proxy = SETTINGS[:katello][:cdn]
+    original_default_global_content_proxy = HttpProxy.default_global_content_proxy&.full_url
+
+    HttpProxy.default_global_content_proxy.full_url = global_content_proxy_mock
+
+    SETTINGS[:katello][:cdn] = katello_cdn_proxy_mock
+
+    # default_global_content_proxy should be selected.
+    assert_equal global_content_proxy_mock, ForemanRhCloud.get_proxy
+
+    HttpProxy.default_global_content_proxy.full_url = nil
+
+    # now cdn_proxy should be selected.
+    assert_equal katello_cdn_proxy_mock, ForemanRhCloud.get_proxy
+
+    SETTINGS[:katello][:cdn] = nil
+
+    # finally global_foreman_proxy should be selected.
+    assert_equal global_foreman_proxy_mock, ForemanRhCloud.get_proxy
```

Comment:
Move this to a separate test method.

You want three tests:

``` ruby
test 'selects global content proxy'
  setup_global_content_proxy
  setup_global_foreman_proxy
  setup_cdn_proxy do
    assert_equal @global_content_proxy_mock, ForemanRhCloud.get_proxy
  end
end

test 'selects cdn proxy'
  setup_global_foreman_proxy
  setup_cdn_proxy do
    assert_equal @katello_cdn_proxy_mock, ForemanRhCloud.get_proxy
  end
end

test 'selects cdn proxy'
  setup_global_foreman_proxy

  assert_equal @global_foreman_proxy_mock, ForemanRhCloud.get_proxy
end

def setup_global_content_proxy
  http_proxy = FactoryBot.create(:http_proxy, url: @global_content_proxy_mock
  HttpProxy.stubs(:default_global_content_proxy).returns(http_proxy)
end

def setup_global_foreman_proxy
    FactoryBot.create(:setting, :name => 'http_proxy', :value => global_foreman_proxy_mock)
end

def setup_cdn_proxy
   old_cdn_setting = SETTINGS[:katello][:cdn]
   SETTINGS[:katello][:cdn] = katello_cdn_proxy_mock
   yield
ensure
  SETTINGS[:katello][:cdn] = old_cdn_setting
end
```

Reply:
Wow thanks!! it really helped :)